lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+#include "FreeRTOS.h"
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
+#include "compartment.h"
+#include "tcpip-internal.h"
+#include <atomic>
+#include <cheri.hh>
+#include <debug.hh>
+#include <locks.hh>
+#include <priv/riscv.h>
+#include <simulator.h>
+#include <vector>
+
+using DebugErrorHandler = ConditionalDebug<true, "TCP/IP Stack error handler">;
+using CHERI::Capability;
+
+// Global state to update as part of the reset.
+extern std::atomic<uint32_t> currentSocketEpoch;
+extern std::atomic<uint8_t>  userThreadCount;
+
+// Locks and futexes to release as part of the reset
+extern struct FlagLockState       ipThreadLockState;
+extern struct RecursiveMutexState __CriticalSectionFlagLock;
+extern struct RecursiveMutexState __SuspendFlagLock;
+extern QueueHandle_t              xNetworkEventQueue;
+
+// APIs to call as part of the reset
+extern void free_buffer_manager_memory();
+extern void network_restart();
+
+/// Thread ID of the network thread.
+extern uint16_t networkThreadID;
+
+/**
+ * Free the compartment's heap memory.
+ *
+ * Note that socket memory will not be freed because sockets are allocated with
+ * user-passed capabilities which we do not store. API users are supposed to
+ * close them, which will trigger a free.
+ */
+__always_inline static inline void free_compartment_memory()
+{
+	// Global heap capability.
+	heap_free_all(MALLOC_CAPABILITY);
+	// Buffer manager capability. If the buffer manager is using the global
+	// heap capability, this will do nothing.
+	free_buffer_manager_memory();
+}
+
+/**
+ * Reset the network stack state.
+ *
+ * We go through all locks used in the TCP/IP compartment and set them for
+ * destruction. The list of synchronization primitives resetted here was
+ * extracted through a manual study of the compartment's code-base: this may
+ * therefore break if new releases of FreeRTOS+TCP introduce new locks. In the
+ * future, we may want to come up with a more systematic approach.
+ *
+ * This function is designed to be robust against most types of compartment
+ * corruption, however we do assume that:
+ * - 'reset-critical' data has not been corrupted
+ * - the control-flow of threads in the compartment has not been altered
+ */
+extern "C" void reset_network_stack_state()
+{
+	const bool isUserThread = thread_id_get() != networkThreadID;
+	const bool isIpThread   = !isUserThread;
+
+	if (isUserThread)
+	{
+		DebugErrorHandler::log(
+		  "User thread TCP/IP stack error handler called!");
+		userThreadCount--;
+	}
+	else
+	{
+		DebugErrorHandler::log(
+		  "Network thread TCP/IP stack error handler called!");
+	}
+
+	// Manually unlock the sealed sockets lock list if it was held.
```

Comment:
Under what situations can this happen?  My expectation is that it's only if it crashes while adding to the linked list.  Is the list state well defined in that case?  Are operations correctly serialised with compiler barriers to ensure that the list is at least traversable in one direction?

Reply:
Not necessarily. The thread can also crash in `network_socket_create_and_bind` while holding the lock (but not editing the list) because we hold the lock for a bit longer to simplify error handling.

We don't guarantee reset if the list is corrupted (it is "reset critical").

I added comments to clarify both of this.